### PR TITLE
Allow Javadoc jars to be created

### PR DESCRIPTION
**Pull request by shs96c**

### CODE REVIEW

**Comment by shs96c** [comment]
This includes some of the building blocks for the maven work. Notably, there's `JarShape` which makes it easier to handle über jars and the like.

**Comment by shs96c** [comment]
I need to make the Javadoc class a MavenPiblishable. Will try and get that updated today.

**Comment by shs96c** [comment]
Ready for review :)

File: src/com/facebook/buck/jvm/java/JarShape.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public enum JarShape {
+  MAVEN {
+    @Override
+    public Summary gatherDeps(BuildRule root) {
+      // Our life is a lot easier if the root is a JavaLibrary :)
+      if (!(root instanceof JavaLibrary)) {
+        throw new RuntimeException("Boom");
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/JarShape.java]: How about something that doesn't result in the user ducking for cover?

File: src/com/facebook/buck/jvm/java/JarShape.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    }
+  },
+  SINGLE {
+    @Override
+    public Summary gatherDeps(BuildRule root) {
+      ImmutableSet<JavaLibrary> classpath;
+      if (root instanceof HasClasspathEntries) {
+        classpath = ((HasClasspathEntries) root).getTransitiveClasspathDeps();
+      } else {
+        classpath = ImmutableSet.of();
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/JarShape.java]: should we error out for things that don't have a classpath?

File: src/com/facebook/buck/jvm/java/JarShape.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      Set<HasMavenCoordinates> mavenDeps = new HashSet<>();
+      Set<JavaLibrary> combinedTransitiveDeps = new HashSet<>();
+      new AbstractBreadthFirstTraversal<JavaLibrary>((JavaLibrary) root) {
+        @Override
+        public Iterable<JavaLibrary> visit(JavaLibrary javaLibrary) throws RuntimeException {
+          if (!javaLibrary.equals(root) && javaLibrary.getMavenCoords().isPresent()) {
+            mavenDeps.add(javaLibrary);
+            // We may have a transitive dep on another maven dep.
+            combinedTransitiveDeps.addAll(
+                javaLibrary.getTransitiveClasspathDeps().stream()
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/JarShape.java]: this, combined with the traversal, is O(n^2). How hard would it be to make it more efficient?

File: src/com/facebook/buck/jvm/java/JarShape.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      if (root instanceof HasClasspathEntries) {
+        HasClasspathEntries entries = (HasClasspathEntries) root;
+
+        return new Summary(
+            entries.getTransitiveClasspathDeps(),
+            entries.getTransitiveClasspathDeps(),
+            ImmutableSortedSet.of());
+      }
+
+      throw new RuntimeException("Not handled yet");
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/JarShape.java]: marginally better than "Boom!" but equally helpful :)

File: src/com/facebook/buck/jvm/java/Javadoc.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+      BuildableContext buildableContext) {
+    ImmutableList.Builder<Step> steps = ImmutableList.builder();
+
+    steps.add(new MkdirStep(getProjectFilesystem(), output.getParent()));
+    steps.add(new RmStep(getProjectFilesystem(), output, /* force deletion */ true));
+
+    if (sources.isEmpty()) {
+      // Create an empty jar. This keeps things that want a jar happy without causing javadoc to
+      // choke on there not being any sources.
+      steps.add(
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/Javadoc.java]: please express the control flow such that there is only one call to `new ZipStep`

File: src/com/facebook/buck/jvm/java/Javadoc.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+          new ZipStep(
+              getProjectFilesystem(),
+              output,
+              ImmutableSet.<Path>of(),
+              /* junk paths */ false,
+              ZipCompressionLevel.MIN_COMPRESSION_LEVEL,
+              output));
+      return steps.build();
+    } else {
+      Path atFilePath = scratchDir.resolve("all-sources.txt");
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/Javadoc.java]: atFilePath -> sourcesListFilePath ?

File: src/com/facebook/buck/jvm/java/JarShape.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        classpath = ImmutableSet.of();
+      }
+
+      return new Summary(
+          ImmutableSortedSet.of(root),
+          classpath,
+          ImmutableSortedSet.of());
+    }
+  },
+  UBER {
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/JarShape.java]: This is not even used yet, right? Could we add it in the commit that actually needs it?

File: src/com/facebook/buck/jvm/java/JarShape.java (hunk 1 above)
Comment (shs96c) [inline src/com/facebook/buck/jvm/java/JarShape.java]: Done.

File: src/com/facebook/buck/jvm/java/JarShape.java (hunk 3 above)
Comment (shs96c) [inline src/com/facebook/buck/jvm/java/JarShape.java]: In Buck and Selenium, running the maven target for the complete source tree is still pretty fast --- it takes longer to create the javadocs by a long shot. Couple this with the fact that the publish command can only publish a single target at a time, and I think that the clarity of this code is okay for now. We can always ask @yiding to do some magic ;)

File: src/com/facebook/buck/jvm/java/JarShape.java (hunk 7 above)
Comment (shs96c) [inline src/com/facebook/buck/jvm/java/JarShape.java]: Done.

File: src/com/facebook/buck/jvm/java/Javadoc.java (hunk 6 above)
Comment (shs96c) [inline src/com/facebook/buck/jvm/java/Javadoc.java]: Done

File: src/com/facebook/buck/jvm/java/JarShape.java (hunk 3 above)
Comment (shs96c) [inline src/com/facebook/buck/jvm/java/JarShape.java]: But now sorted out with something that's easier to read too.

File: src/com/facebook/buck/jvm/java/Javadoc.java (hunk 5 above)
Comment (shs96c) [inline src/com/facebook/buck/jvm/java/Javadoc.java]: I'm not sure that makes the code more readable. We basically have the fast-path of doing nothing, or something that takes legwork. The `else` keyword isn't necessary, though.

File: src/com/facebook/buck/jvm/java/JarShape.java [hunk 8]
```diff
@@ -35,36 +34,30 @@ public Summary gatherDeps(BuildRule root) {
...
+        mavenDeps.removeAll(
+            transitive.stream()
+                .filter(dep -> !dep.equals(mavenDep))
+                .collect(Collectors.toSet()));
+      }
+
       return new Summary(
           toPackage,
-          classpath,
+          ((HasClasspathEntries) root).getTransitiveClasspathDeps(),
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/JarShape.java]: you call this twice, maybe extract to a variable?

Comment (shs96c) [inline src/com/facebook/buck/jvm/java/JarShape.java]: Done

**Comment by marcinkosiba** [comment]
fix `ant` build and checkstyle (see travis failures)

**Pull request closed without merging**